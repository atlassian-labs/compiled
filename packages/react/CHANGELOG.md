# @compiled/react

## 0.10.2

### Patch Changes

- 254a6f6: Added ESLint rule to prevent use of extraneous packages, and added these usages of these packages as dependencies. Added new namespace `@compiled-private` to prevent name clashes with existing npm packages.
- c757259: Update type definition dependencies
- 6649528: Changed the SSR check to be based on the presence of `document` instead of looking for Node processes.

## 0.10.1

### Patch Changes

- d3e257c: Fixes `css` function types to allow nested styles
- 8c9ab8c: Update `homepage` and other `package.json` properties
- 8c9ab8c: JSDoc descriptions for all exports have been updated, let us know what you think!
- 8c9ab8c: `ClassNames` component now has its style prop typed as `CSSProperties` instead of a string map

## 0.10.0

### Minor Changes

- 427cead: **Breaking change:** When using the `css` prop with [TypeScript](https://www.typescriptlang.org) you now need to declare a JSX pragma enabling types for that module. Previously when importing the `@compiled/react` package the global JSX namespace would be polluted as a side effect potentially causing collisions with other CSS-in-JS libraries. Now thanks to the use of [locally scoped JSX namespaces](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html#locally-scoped-jsx-namespaces) the global JSX namespace will no longer be polluted.

  As an added bonus the `css` prop will only be available on JSX elements that have also defined a `className` prop with the potential for more type safe features later on.

  Make sure to update all Compiled dependencies to latest when adopting this change.

  **Automatic runtime**

  ```diff
  -import '@compiled/react';
  +/** @jsxImportSource @compiled/react */

  <div css={{ display: 'block' }} />;
  ```

  **Classic runtime**

  ```diff
  -import '@compiled/react';
  +/** @jsx jsx */
  +import { jsx } from '@compiled/react';

  <div css={{ display: 'block' }} />;
  ```

  To aid consumers in adopting this change easily, a new ESLint rule `jsx-pragma` has been created which will automatically migrate you to use a JSX pragma if missing when running with `--fix`. The rule takes an option to configure the runtime (either classic or automatic) and defaults to automatic.

  ```sh
  npm i @compiled/eslint-plugin
  ```

  ```json
  {
    "rules": {
      "@compiled/jsx-pragma": ["error", { "runtime": "classic" }]
    }
  }
  ```

### Patch Changes

- 79cfb08: Internal refactor changing how the TypeScript compiler picks up source files.

## 0.9.1

### Patch Changes

- 4309aaa: Patch inexact flow type on styled

## 0.9.0

### Minor Changes

- 2092839: Allow inline strings and inline css mixins in conditional expressions. Fix ordering of styles in template literals.

## 0.8.0

### Minor Changes

- 4210ff6: Add flow types support
- 53935b3: Add `ObjectExpression` support to `css`

## 0.7.0

### Minor Changes

- bcb2a68: Add support for `keyframes`
- a7ab8e1: Add support for conditional rules for `Styled`

## 0.6.13

### Patch Changes

- 13c3a60: add support of additional parameters to css function

## 0.6.12

### Patch Changes

- b5b4e8a: Catch unhandled exception on inserting rules with prefixed selectors.

## 0.6.11

### Patch Changes

- ee3363e: Fix HTML characters escapes in style tags on SSR.

## 0.6.10

### Patch Changes

- 40bc0d9: Package descriptions have been updated.
- 1b1c964: The `css` mixin API is now available,
  functioning similarly to the [emotion equivalent](https://emotion.sh/docs/composition).

  ```jsx
  import { css } from '@compiled/react';

  <div
    css={css`
      display: flex;
      font-size: 50px;
      color: blue;
    `}
  >
    blue text
  </div>;
  ```

  For more help, read the docs: https://compiledcssinjs.com/docs/css.

## 0.6.9

### Patch Changes

- 4032cd4: Memo has been removed from the style component which was breaking re-renders at times.

## 0.6.8

### Patch Changes

- aea3504: Packages now released with [changesets](https://github.com/atlassian/changesets).
