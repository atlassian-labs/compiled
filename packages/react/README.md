# @compiled/react

The primary entrypoint package for Compiled.

## Installation

```bash
npm i @compiled/react
```

## Usage

Detailed docs and example usage can be [found on the documentation website](https://compiledcssinjs.com/docs/pkg-react).
