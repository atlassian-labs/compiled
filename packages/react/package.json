{
  "name": "@compiled/react",
  "version": "0.10.2",
  "description": "A familiar and performant compile time CSS-in-JS library for React.",
  "keywords": [
    "compiled",
    "css-in-js",
    "styled-components",
    "typescript"
  ],
  "homepage": "https://compiledcssinjs.com/docs/pkg-react",
  "bugs": "https://github.com/atlassian-labs/compiled/issues/new?assignees=&labels=bug&template=bug_report.md",
  "repository": {
    "type": "git",
    "url": "https://github.com/atlassian-labs/compiled.git",
    "directory": "packages/react"
  },
  "license": "Apache-2.0",
  "author": "Michael Dougall",
  "sideEffects": false,
  "exports": {
    ".": {
      "import": [
        "./dist/esm/index.js",
        "./src/index.tsx"
      ],
      "require": [
        "./dist/cjs/index.js",
        "./src/index.tsx"
      ]
    },
    "./runtime": {
      "import": [
        "./dist/esm/runtime.js",
        "./src/runtime.tsx"
      ],
      "require": [
        "./dist/cjs/runtime.js",
        "./src/runtime.tsx"
      ]
    },
    "./jsx-runtime": {
      "import": [
        "./dist/esm/jsx/jsx-runtime.js",
        "./src/jsx/jsx-runtime.tsx"
      ],
      "require": [
        "./dist/cjs/jsx/jsx-runtime.js",
        "./src/jsx/jsx-runtime.tsx"
      ]
    },
    "./jsx-dev-runtime": {
      "import": [
        "./dist/esm/jsx/jsx-dev-runtime.js",
        "./src/jsx/jsx-dev-runtime.tsx"
      ],
      "require": [
        "./dist/cjs/jsx/jsx-dev-runtime.js",
        "./src/jsx/jsx-dev-runtime.tsx"
      ]
    }
  },
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.js",
  "browser": "./dist/browser/index.js",
  "types": "./dist/esm/index.d.ts",
  "files": [
    "dist",
    "src",
    "runtime",
    "jsx-runtime",
    "jsx-dev-runtime"
  ],
  "dependencies": {
    "csstype": "^3.0.10"
  },
  "devDependencies": {
    "@compiled/benchmark": "^1.0.1",
    "@testing-library/react": "^12.1.2",
    "@types/react-dom": "^17.0.11",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  },
  "peerDependencies": {
    "react": ">= 16.12.0"
  }
}
