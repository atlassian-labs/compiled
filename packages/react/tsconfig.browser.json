{
  "extends": "./tsconfig.options.json",
  "compilerOptions": {
    "outDir": "dist/browser",
    "plugins": [
      {
        "transform": "ts-transform-define",
        "replace": {
          "isNodeEnvironment()": "false"
        }
      }
    ],
    "tsBuildInfoFile": "tsconfig.browser.tsbuildinfo"
  },
  "references": [{ "path": "../benchmark" }, { "path": "../jest" }]
}
