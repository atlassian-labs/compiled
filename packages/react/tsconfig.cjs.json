{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "dist/cjs",
    "module": "commonjs",
    "tsBuildInfoFile": "tsconfig.cjs.tsbuildinfo"
  },
  "references": []
}
