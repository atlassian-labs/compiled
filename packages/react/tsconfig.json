{
  "extends": "../tsconfig.options.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "dist/esm",
    "baseUrl": "src",
    "paths": {
      "@compiled/react": ["index.tsx"]
    },
    "plugins": [
      {
        "transform": "ts-transform-define",
        "replace": {
          "isNodeEnvironment()": "process.env.IS_NODE_EXPRESSION"
        }
      }
    ],
    "tsBuildInfoFile": "tsconfig.esm.tsbuildinfo"
  },
  "references": []
}
