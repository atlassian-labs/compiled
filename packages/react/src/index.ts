import { createElement } from 'react';

import type { CompiledJSX } from './jsx/jsx-local-namespace';

export { ClassNames } from './class-names';
export { createStrictAPI } from './create-strict-api';
export { default as css } from './css';
export { default as cssMap } from './css-map';
export { keyframes } from './keyframes';
export { styled } from './styled';
export type {
  CSSProperties,
  CSSProps,
  CSSPseudos,
  CssFunction,
  CssType,
  StrictCSSProperties,
} from './types';
export { type XCSSAllProperties, type XCSSAllPseudos, type XCSSProp, cx } from './xcss-prop';

// Pass through the (classic) jsx runtime.
// Compiled currently doesn't define its own and uses this purely to enable a local jsx namespace.
export const jsx = createElement;

export namespace jsx {
  export namespace JSX {
    export type Element = CompiledJSX.Element;
    export type ElementClass = CompiledJSX.ElementClass;
    export type ElementAttributesProperty = CompiledJSX.ElementAttributesProperty;
    export type ElementChildrenAttribute = CompiledJSX.ElementChildrenAttribute;
    export type LibraryManagedAttributes<C, P> = CompiledJSX.LibraryManagedAttributes<C, P>;
    export type IntrinsicAttributes = CompiledJSX.IntrinsicAttributes;
    export type IntrinsicClassAttributes<T> = CompiledJSX.IntrinsicClassAttributes<T>;
    export type IntrinsicElements = CompiledJSX.IntrinsicElements;
  }
}
