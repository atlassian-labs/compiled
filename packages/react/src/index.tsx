import { createElement } from 'react';

import type { CompiledJSX } from './jsx/jsx-local-namespace';
import type { CssFunction, CSSProps } from './types';

export { keyframes } from './keyframes';
export { styled } from './styled';
export { ClassNames } from './class-names';
export { default as css } from './css';

// Pass through the (classic) jsx runtime.
// Compiled currently doesn't define its own and uses this purely to enable a local jsx namespace.
export const jsx = createElement;

export type { CssFunction, CSSProps };
export type { CssObject } from './styled';

export namespace jsx {
  export namespace JSX {
    export type Element = CompiledJSX.Element;
    export type ElementClass = CompiledJSX.ElementClass;
    export type ElementAttributesProperty = CompiledJSX.ElementAttributesProperty;
    export type ElementChildrenAttribute = CompiledJSX.ElementChildrenAttribute;
    export type LibraryManagedAttributes<C, P> = CompiledJSX.LibraryManagedAttributes<C, P>;
    export type IntrinsicAttributes = CompiledJSX.IntrinsicAttributes;
    export type IntrinsicClassAttributes<T> = CompiledJSX.IntrinsicClassAttributes<T>;
    export type IntrinsicElements = CompiledJSX.IntrinsicElements;
  }
}
