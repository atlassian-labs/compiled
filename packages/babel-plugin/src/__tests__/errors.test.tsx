import { transformSync } from '@babel/core';
import babelPlugin from '../index';

const transform = (code: string) => {
  return transformSync(code, {
    configFile: false,
    babelrc: false,
    compact: true,
    plugins: [babelPlugin],
    // Turn off code highlighting so snapshots are human readable.
    highlightCode: false,
  })?.code;
};

describe('error handling', () => {
  it('should throw when using using an invalid css node', () => {
    expect(() => {
      transform(`
      import '@compiled/react';
      import React from 'react';

      <div css={() => {}} />
    `);
    }).toThrowErrorMatchingSnapshot();
  });

  it('should throw when spreading an identifier that does not exist', () => {
    expect(() => {
      transform(`
      import '@compiled/react';
      import React from 'react';

      <div css={{ ...dontexist }} />
    `);
    }).toThrowErrorMatchingSnapshot();
  });

  it('should throw when referencing an identifier that does not exist', () => {
    expect(() => {
      transform(`
      import '@compiled/react';
      import React from 'react';

      <div css={dontexist} />
    `);
    }).toThrowErrorMatchingSnapshot();
  });

  it('should throw when referencing an identifier that isnt supported', () => {
    expect(() => {
      transform(`
      import '@compiled/react';
      import React from 'react';

      class HelloWorld {}

      <div css={HelloWorld} />
    `);
    }).toThrowErrorMatchingSnapshot();
  });

  it('should throw when composing invalid css', () => {
    expect(() => {
      transform(`
      import '@compiled/react';
      import React from 'react';

      <div css={[...hello]} />
    `);
    }).toThrowErrorMatchingSnapshot();
  });
});
