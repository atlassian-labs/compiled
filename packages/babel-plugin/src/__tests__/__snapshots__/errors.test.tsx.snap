// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error handling should throw when composing invalid css 1`] = `
"unknown: SpreadElement isn't a supported CSS type - try using an object or string (5:16).
  3 |       import React from 'react';
  4 | 
> 5 |       <div css={[...hello]} />
    |                 ^^^^^^^^^^
  6 |     "
`;

exports[`error handling should throw when referencing an identifier that does not exist 1`] = `
"unknown: Variable could not be found (5:16).
  3 |       import React from 'react';
  4 | 
> 5 |       <div css={dontexist} />
    |                 ^^^^^^^^^
  6 |     "
`;

exports[`error handling should throw when referencing an identifier that isnt supported 1`] = `
"unknown: ClassDeclaration isn't a supported CSS type - try using an object or string (7:16).
  5 |       class HelloWorld {}
  6 | 
> 7 |       <div css={HelloWorld} />
    |                 ^^^^^^^^^^
  8 |     "
`;

exports[`error handling should throw when spreading an identifier that does not exist 1`] = `
"unknown: Variable could not be found (5:21).
  3 |       import React from 'react';
  4 | 
> 5 |       <div css={{ ...dontexist }} />
    |                      ^^^^^^^^^
  6 |     "
`;

exports[`error handling should throw when using using an invalid css node 1`] = `
"unknown: ArrowFunctionExpression isn't a supported CSS type - try using an object or string (5:16).
  3 |       import React from 'react';
  4 | 
> 5 |       <div css={() => {}} />
    |                 ^^^^^^^^
  6 |     "
`;
